### PR TITLE
[flang][NFC] Extract FIROpConversion to its own files

### PR DESCRIPTION
**Pull request by clementval**
This PR extracts `FIROpConversion` and `FIROpAndTypeConversion` templated base patterns to a header file. All the functions from FIROpConversion that do not require the template argument are moved to a base class named `ConvertFIRToLLVMPattern`.
This move is done so the `FIROpConversion` pattern and all its utility functions can be reused outside of the codegen pass.

For the most part the code is only moved to the new files and not modified. The only update is that addition of the PatternBenefit argument with a default value to the constructor so it can be forwarded to the `ConversionPattern` ctor. 

This split is done in a similar way for the `ConvertOpToLLVMPattern` base pattern that is based on the `ConvertToLLVMPattern` base class in `mlir/include/mlir/Conversion/LLVMCommon/Pattern.h`.

### CODE REVIEW

File: flang/include/flang/Optimizer/CodeGen/FIROpPatterns.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+#include "flang/Optimizer/CodeGen/TypeConverter.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+
+namespace fir {
+
+struct FIRToLLVMPassOptions;
+
+// TODO: This should really be recovered from the specified target.
```
Comment (jeanPerier) [inline flang/include/flang/Optimizer/CodeGen/FIROpPatterns.h:20]: I am not sure the default address space is target dependent (maybe it is defined in some LLVM header though). I think the TODO comment applied to the default alignment.

**Review by jeanPerier — APPROVED** [review]
Thanks for this cleanup!

File: flang/include/flang/Optimizer/CodeGen/FIROpPatterns.h (hunk 1 above)
Comment (vzakhari) [reply flang/include/flang/Optimizer/CodeGen/FIROpPatterns.h:20]: Right, `0` is always the default address space in LLVM.  Address spaces for specific objects (e.g. global variables) may be dependent on the target.

Comment (clementval) [reply flang/include/flang/Optimizer/CodeGen/FIROpPatterns.h:20]: I copied blindly the comment from the file and you are right that this does not apply for this line. I'll remove it.

**Pull request merged**